brilift/src/main.rs

```diff
@@ -318,168 +353,175 @@ fn gen_binary(
 }
 
 /// An environment for translating Bril into CLIF.
-struct CompileEnv<'a> {
+struct CompileEnv<'a, M: Module> {
+    module: &'a mut M,
```

Comment:
I'm actually a little wary of putting the `Module` into this environment. Part of the idea with separating `compile_inst` from the `Translator` structs was to _divorce_ it from the underlying module as much as possible. The reason is the same reason why Cranelift separates the core `cranelift_codegen` crate from `cranelift_module`: it's drawing a line between the intraprocedural "let's generate raw machine code sequences" functionality from the broader interprocedural context of stitching together multiple functions into a complete module. So despite the secondary goal of lazily only using the necessary runtime functions in a given Bril function, I would love to preserve this separation as much as possible…

Reply:
I think I could pass a closure which hides away any underlying implementation of how from a given `RTFunc`/`String`, an `ir::FuncRef` is produced. This way you could do either method, using an EnumMap/HashMap or a function lookup, and from the `CompileEnv`'s perspective it is none the wiser. I think that encapsulates the separation you are looking for but maybe it's too much indirection.

Reply:
FWIW, the alternative idea I was pondering for importing only the used functions was to do a second pass over the instructions… that is, an earlier pass could look for all functions referenced by any instruction and import those, all up front.